jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/SubscribeSnsDialog.kt

```diff
@@ -0,0 +1,95 @@
+
+import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.DialogWrapper
+import com.intellij.openapi.ui.ValidationInfo
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import org.jetbrains.annotations.TestOnly
+import software.amazon.awssdk.services.sns.SnsClient
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.core.utils.warn
+import software.aws.toolkits.jetbrains.core.awsClient
+import software.aws.toolkits.jetbrains.core.explorer.refreshAwsTree
+import software.aws.toolkits.jetbrains.services.sqs.resources.SqsResources
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.resources.message
+import javax.swing.JComponent
+
+private val NOTIFICATION_TITLE = message("sqs.service_name")
+
+class SubscribeSnsDialog(
+    private val project: Project,
+    private val queue: Queue
+) : DialogWrapper(project), CoroutineScope by ApplicationThreadPoolScope("SubscribeSnsDialog") {
+    private val snsClient: SnsClient = project.awsClient()
+    val view = SubscribeSnsPanel(project)
+
+    init {
+        title = message("sqs.subscribe.sns")
+        setOKButtonText(message("sqs.subscribe.sns.subscribe"))
+
+        init()
+    }
+
+    override fun createCenterPanel(): JComponent? = view.component
+
+    override fun getPreferredFocusedComponent(): JComponent? = view.topicSelector
+
+    override fun doValidate(): ValidationInfo? {
+        if (topicSelected().isEmpty()) {
+            return ValidationInfo(message("sqs.subscribe.sns.validation.empty_topic"), view.topicSelector)
+        }
+        return null
+    }
+
+    override fun doOKAction() {
+        if (isOKActionEnabled) {
+            setOKButtonText(message("sqs.subscribe.sns.in_progress"))
+            isOKActionEnabled = false
+
+            launch {
+                try {
+                    subscribe(topicSelected())
+                    runInEdt(ModalityState.any()) {
+                        close(OK_EXIT_CODE)
+                    }
+                    project.refreshAwsTree(SqsResources.LIST_QUEUE_URLS)
+                    notifyInfo(NOTIFICATION_TITLE, message("sqs.subscribe.sns.success", topicSelected()), project)
+                } catch (e: Exception) {
+                    LOG.warn(e) { message("sqs.subscribe.sns.failed", queue.queueName, topicSelected()) }
+                    setErrorText(e.message)
+                    setOKButtonText(message("sqs.subscribe.sns.subscribe"))
+                    isOKActionEnabled = true
+                }
+            }
+        }
+    }
+
+    private fun topicSelected(): String =
+        if (view.topicSelector.selected() == null) {
+            ""
+        } else {
+            view.topicSelector.selected()!!.arn
+        }
+
+    @TestOnly
```

Comment:
This isn't test only, marking it `internal` matches what you meant by adding `TestOnly`